The binary type is abnormal during the data migration data check phase

The binary type is abnormal during the data migration data check phase
### Which version of ShardingSphere did you use?
master branch
MySQL5.6,5.7
### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
Sharding-Scalinlg
### Expected behavior
Successful migration
### Actual behavior
Migration data check phase failed
### Reason analyze (If you can)
The binary type found in source is Blob, and the type found in target is byte[]
Possible problem location：
`org.apache.shardingsphere.data.pipeline.core.check.consistency.DataConsistencyCheckerImpl#checkRecordsContent`
### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
```sql
-- create table
create table table_string
(
    id          bigint 
        primary key,
    shard_id    bigint               not null,
    t_char      char                 null,
    t_varchar   varchar(100)         null,
    t_binary    binary(1)            null,
    t_varbinary varbinary(100)       null,
    t_blob      blob                 null,
    t_text      text                 null,
    t_enum      enum ('1', '2', '3') null,
    t_set       set ('1', '2', '3')  null
);
-- insert datas
insert into table_string (id, shard_id, t_char, t_varchar,
                            t_binary, t_varbinary,
                            t_blob, t_text, t_enum, t_set)
VALUES (1, 2, 'n', 'nihao shijie', 0, '0101110', 'testtesttesttest\stesttesttest', 'testtesttesttest', '1', '1'),
       (2, 1, 'i', 'nihao shijie', 1, '0101110', 'testtesttesttest\stesttesttest', 'testtesttesttest', '2', '3'),
       (3, 3, 'h', 'nihao shijie', 1, '0101110', 'testtesttesttest\stesttesttest', 'testtesttesttest', '3', '1'),
       (4, 6, 'a', 'nihao shijie', 0, '0101110', 'testtesttesttest\stesttesttest', 'testtesttesttest', '2', '1'),
       (5, 5, 'o', 'nihao shijie', 1, '0101110', 'testtesttesttest\stesttesttest', 'testtesttesttest', '3', '3'),
       (6, 4, 'b', 'nihao shijie', 0, '0101110', 'testtesttesttest\stesttesttest', 'testtesttesttest', '3', '2');

-- add resource
ADD RESOURCE ds_2 (
    URL="jdbc:mysql://127.0.0.1:3306/demo_ds_2?serverTimezone=UTC&useSSL=false",
    USER=root,
    PASSWORD=123456,
    PROPERTIES("maximumPoolSize"=50,"idleTimeout"="60000")
), ds_3 (
    URL="jdbc:mysql://127.0.0.1:3306/demo_ds_3?serverTimezone=UTC&useSSL=false",
    USER=root,
    PASSWORD=123456,
    PROPERTIES("maximumPoolSize"=50,"idleTimeout"="60000")
);
-- start scaling
ALTER SHARDING TABLE RULE table_string (
    RESOURCES(ds_2,ds_3),
    SHARDING_COLUMN=id,
    TYPE(NAME=HASH_MOD,PROPERTIES("sharding-count"=2)),
    GENERATED_KEY(COLUMN=id,TYPE(NAME=snowflake,PROPERTIES("worker-id"=123)))
);
```
data check
```java
[INFO ] 2022-01-07 18:03:00.045 [_finished_check_Worker-1] o.a.s.d.p.c.a.i.PipelineJobAPIImpl - Data consistency check for job 401487516776231
[INFO ] 2022-01-07 18:03:00.053 [_finished_check_Worker-1] o.a.s.s.s.ShardingRuleAlteredJobConfigurationPreparer - toTaskConfigs, dataSourceName=ds_0, taskConfig=TaskConfiguration(handleConfig=HandleConfiguration(jobId=401487516776231, concurrency=3, retryTimes=3, tablesFirstDataNodes=table_string:ds_0.table_string_0, jobShardingDataNodes=[table_string:ds_0.table_string_0,ds_0.table_string_1, table_string:ds_1.table_string_0,ds_1.table_string_1], logicTables=table_string, jobShardingItem=0, shardingSize=10000000, sourceDatabaseType=MySQL, targetDatabaseType=MySQL), dumperConfig=DumperConfiguration(dataSourceName=ds_0, blockQueueSize=10000, position=null, tableNameMap={table_string_0=table_string, table_string_1=table_string}), importerConfig=ImporterConfiguration(shardingColumnsMap={table_string=[id]}, retryTimes=3))
[INFO ] 2022-01-07 18:03:00.392 [_finished_check_Worker-1] o.a.s.d.p.c.a.i.PipelineJobAPIImpl - Scaling job 401487516776231 with check algorithm 'org.apache.shardingsphere.data.pipeline.core.spi.check.consistency.DataMatchDataConsistencyCheckAlgorithm' data consistency checker result {table_string=DataConsistencyCheckResult(sourceRecordsCount=6, targetRecordsCount=6, recordsCountMatched=true, recordsContentMatched=false)}
[ERROR] 2022-01-07 18:03:00.392 [_finished_check_Worker-1] o.a.s.d.p.c.a.i.PipelineJobAPIImpl - Scaling job: 401487516776231, table: table_string data consistency check failed, recordsContentMatched: false, recordsCountMatched: true
[INFO ] 2022-01-07 18:03:00.392 [_finished_check_Worker-1] o.a.s.d.p.c.a.i.GovernanceRepositoryAPIImpl - persist job check result 'false' for job 401487516776231
[ERROR] 2022-01-07 18:03:00.394 [_finished_check_Worker-1] o.a.s.d.p.c.a.i.PipelineJobAPIImpl - Scaling job: 401487516776231, table: table_string data consistency check failed, recordsContentMatched: false, recordsCountMatched: true
[ERROR] 2022-01-07 18:03:00.394 [_finished_check_Worker-1] o.a.s.d.p.core.job.FinishedCheckJob - data consistency check failed, job 401487516776231
```
### Example codes for reproduce this issue (such as a github link).

--- a/shardingsphere-infra/shardingsphere-infra-executor/src/main/java/org/apache/shardingsphere/infra/executor/sql/execute/result/query/impl/driver/jdbc/type/memory/JDBCRowsLoader.java
+++ b/shardingsphere-infra/shardingsphere-infra-executor/src/main/java/org/apache/shardingsphere/infra/executor/sql/execute/result/query/impl/driver/jdbc/type/memory/JDBCRowsLoader.java
@@ -97,10 +97,11 @@ public final class JDBCRowsLoader {
             case Types.CLOB:
                 return resultSet.getClob(columnIndex);
             case Types.BLOB:
+                return resultSet.getBlob(columnIndex);
             case Types.BINARY:
             case Types.VARBINARY:
             case Types.LONGVARBINARY:
-                return resultSet.getBlob(columnIndex);
+                return resultSet.getBytes(columnIndex);
             case Types.ARRAY:
                 return resultSet.getArray(columnIndex);
             default:

--- a/shardingsphere-infra/shardingsphere-infra-executor/src/test/java/org/apache/shardingsphere/infra/executor/sql/execute/result/query/impl/driver/jdbc/type/memory/JDBCMemoryQueryResultTest.java
+++ b/shardingsphere-infra/shardingsphere-infra-executor/src/test/java/org/apache/shardingsphere/infra/executor/sql/execute/result/query/impl/driver/jdbc/type/memory/JDBCMemoryQueryResultTest.java
@@ -276,33 +276,33 @@ public final class JDBCMemoryQueryResultTest {
     @Test
     public void assertGetValueByBinary() throws SQLException {
         ResultSet resultSet = getMockedResultSet(Types.BINARY);
-        Blob value = mock(Blob.class);
-        when(resultSet.getBlob(1)).thenReturn(value);
+        byte[] value = new byte[10];
+        when(resultSet.getBytes(1)).thenReturn(value);
         JDBCMemoryQueryResult actual = new JDBCMemoryQueryResult(resultSet);
         assertTrue(actual.next());
-        assertThat(actual.getValue(1, Blob.class), is(value));
+        assertThat(actual.getValue(1, byte[].class), is(value));
         assertFalse(actual.next());
     }
     
     @Test
     public void assertGetValueByVarBinary() throws SQLException {
         ResultSet resultSet = getMockedResultSet(Types.VARBINARY);
-        Blob value = mock(Blob.class);
-        when(resultSet.getBlob(1)).thenReturn(value);
+        byte[] value = new byte[10];
+        when(resultSet.getBytes(1)).thenReturn(value);
         JDBCMemoryQueryResult actual = new JDBCMemoryQueryResult(resultSet);
         assertTrue(actual.next());
-        assertThat(actual.getValue(1, Blob.class), is(value));
+        assertThat(actual.getValue(1, byte[].class), is(value));
         assertFalse(actual.next());
     }
     
     @Test
     public void assertGetValueByLongVarBinary() throws SQLException {
         ResultSet resultSet = getMockedResultSet(Types.LONGVARBINARY);
-        Blob value = mock(Blob.class);
-        when(resultSet.getBlob(1)).thenReturn(value);
+        byte[] value = new byte[10];
+        when(resultSet.getBytes(1)).thenReturn(value);
         JDBCMemoryQueryResult actual = new JDBCMemoryQueryResult(resultSet);
         assertTrue(actual.next());
-        assertThat(actual.getValue(1, Blob.class), is(value));
+        assertThat(actual.getValue(1, byte[].class), is(value));
         assertFalse(actual.next());
     }
     

--- a/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/spi/check/consistency/AbstractStreamingSingleTableDataCalculator.java
+++ b/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/spi/check/consistency/AbstractStreamingSingleTableDataCalculator.java
@@ -82,7 +82,7 @@ public abstract class AbstractStreamingSingleTableDataCalculator extends Abstrac
             Optional<Object> nextResult = this.nextResult;
             dataCalculateParameter.setPreviousCalculatedResult(nextResult.orElse(null));
             this.nextResult = null;
-            return nextResult;
+            return nextResult.orElse(null);
         }
         
         private void calculateIfNecessary() {

--- a/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/spi/check/consistency/DataMatchSingleTableDataCalculator.java
+++ b/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/spi/check/consistency/DataMatchSingleTableDataCalculator.java
@@ -18,11 +18,12 @@
 package org.apache.shardingsphere.data.pipeline.core.spi.check.consistency;
 
 import com.google.common.base.Strings;
-import lombok.EqualsAndHashCode;
 import lombok.Getter;
 import lombok.NonNull;
 import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
 import org.apache.shardingsphere.data.pipeline.api.check.consistency.DataCalculateParameter;
 import org.apache.shardingsphere.data.pipeline.core.exception.PipelineDataConsistencyCheckFailedException;
 import org.apache.shardingsphere.data.pipeline.spi.sqlbuilder.PipelineSQLBuilder;
@@ -37,6 +38,7 @@ import java.sql.ResultSetMetaData;
 import java.sql.SQLException;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Iterator;
 import java.util.Optional;
 import java.util.Properties;
 
@@ -115,7 +117,6 @@ public final class DataMatchSingleTableDataCalculator extends AbstractStreamingS
     
     @RequiredArgsConstructor
     @Getter
-    @EqualsAndHashCode
     private static final class CalculatedResult {
         
         @NonNull
@@ -124,5 +125,39 @@ public final class DataMatchSingleTableDataCalculator extends AbstractStreamingS
         private final int recordCount;
         
         private final Collection<Collection<Object>> records;
+    
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (!(o instanceof CalculatedResult)) {
+                return false;
+            }
+        
+            final CalculatedResult that = (CalculatedResult) o;
+    
+            boolean equalsFirst = new EqualsBuilder().append(getRecordCount(), that.getRecordCount()).append(getMaxUniqueKeyValue(), that.getMaxUniqueKeyValue()).isEquals();
+            if (!equalsFirst) {
+                return false;
+            }
+            Iterator<Collection<Object>> thisIterator = this.records.iterator();
+            Iterator<Collection<Object>> thatIterator = that.records.iterator();
+            while (thisIterator.hasNext() && thatIterator.hasNext()) {
+                Iterator<Object> thisNextIterator = thisIterator.next().iterator();
+                Iterator<Object> thatNextIterator = thatIterator.next().iterator();
+                while (thisNextIterator.hasNext() && thatNextIterator.hasNext()) {
+                    if (!new EqualsBuilder().append(thisNextIterator.next(), thatNextIterator.next()).isEquals()) {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        }
+    
+        @Override
+        public int hashCode() {
+            return new HashCodeBuilder(17, 37).append(getMaxUniqueKeyValue()).append(getRecordCount()).append(getRecords()).toHashCode();
+        }
     }
 }

--- a/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/spi/check/consistency/DataMatchSingleTableDataCalculator.java
+++ b/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/spi/check/consistency/DataMatchSingleTableDataCalculator.java
@@ -141,11 +141,17 @@ public final class DataMatchSingleTableDataCalculator extends AbstractStreamingS
             if (!equalsFirst) {
                 return false;
             }
+            
             Iterator<Collection<Object>> thisIterator = this.records.iterator();
             Iterator<Collection<Object>> thatIterator = that.records.iterator();
             while (thisIterator.hasNext() && thatIterator.hasNext()) {
-                Iterator<Object> thisNextIterator = thisIterator.next().iterator();
-                Iterator<Object> thatNextIterator = thatIterator.next().iterator();
+                Collection<Object> thisNext = thisIterator.next();
+                Collection<Object> thatNext = thatIterator.next();
+                if (thisNext.size() != thatNext.size()) {
+                    return false;
+                }
+                Iterator<Object> thisNextIterator = thisNext.iterator();
+                Iterator<Object> thatNextIterator = thatNext.iterator();
                 while (thisNextIterator.hasNext() && thatNextIterator.hasNext()) {
                     if (!new EqualsBuilder().append(thisNextIterator.next(), thatNextIterator.next()).isEquals()) {
                         return false;

